walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,59 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Arrays;
+
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    if (request.getAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (request.getDuration() > 24 * 60) {
+      return possibleTimes;
+    }
+
+    int earliestPossible = TimeRange.START_OF_DAY;
+
+    for (Event event : events) {
+      if (!isEventConsidered(event, request)) {
+        //ignore event that should not be considered
+        continue;
+      }
+
+      if (earliestPossible < event.getWhen().start()) {
+        addIfEnoughRoom(possibleTimes, TimeRange.fromStartEnd(earliestPossible, event.getWhen().start(), false), request.getDuration());
+      }
+      if (event.getWhen().end() > earliestPossible) {
+        earliestPossible = event.getWhen().end();
+      }
+    }
+
+    if (earliestPossible < TimeRange.END_OF_DAY) {
+      addIfEnoughRoom(possibleTimes, TimeRange.fromStartEnd(earliestPossible, TimeRange.END_OF_DAY, true), request.getDuration());
+    }
+
+    return possibleTimes;
+  }
+
+  private boolean isEventConsidered (Event event, MeetingRequest request) {
+    for (String attendee: event.getAttendees()) {
+      if (request.getAttendees().contains(attendee)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private List<TimeRange> addIfEnoughRoom (List<TimeRange> possibleTimes, TimeRange range, long duration) {
```

Comment:
this function mutate and return. and since this function is simple enough, I think we can use the code directly